renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -836,6 +945,162 @@ VulkanAccelerationStructureManager::Allocation VulkanAccelerationStructureManage
   return readbackmem;
 }
 
+VulkanAccelerationStructureManager::Allocation VulkanAccelerationStructureManager::CreateReplayMemory(
+    MemoryType memType, VkDeviceSize size, VkBufferUsageFlags extraUsageFlags)
+{
+  const VkBufferCreateInfo bufInfo = {
+      VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
+      NULL,
+      0,
+      size,
+      VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT |
+          VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT | extraUsageFlags,
+  };
+
+  const VkDevice d = m_pDriver->GetDev();
+
+  Allocation result;
+  result.size = size;
+
+  VkResult vkr = ObjDisp(d)->CreateBuffer(Unwrap(d), &bufInfo, NULL, &result.buf);
+  CHECK_VKR(m_pDriver, vkr);
+
+  VkMemoryRequirements mrq = {};
+  ObjDisp(d)->GetBufferMemoryRequirements(Unwrap(d), result.buf, &mrq);
+
+  uint32_t memoryTypeIndex = 0;
+  switch(memType)
+  {
+    case MemoryType::Upload:
+      memoryTypeIndex = m_pDriver->GetUploadMemoryIndex(mrq.memoryTypeBits);
+      break;
+    case MemoryType::GPULocal:
+      memoryTypeIndex = m_pDriver->GetGPULocalMemoryIndex(mrq.memoryTypeBits);
+      break;
+    case MemoryType::Readback:
+      memoryTypeIndex = m_pDriver->GetReadbackMemoryIndex(mrq.memoryTypeBits);
+      break;
+  }
+
+  VkMemoryAllocateFlagsInfo flagsInfo = {
+      VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_FLAGS_INFO,
+      NULL,
+      VK_MEMORY_ALLOCATE_DEVICE_ADDRESS_BIT,
+  };
+  VkMemoryAllocateInfo info = {
+      VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
+      &flagsInfo,
+      size,
+      memoryTypeIndex,
+  };
+
+  vkr = ObjDisp(d)->AllocateMemory(Unwrap(d), &info, NULL, &result.mem);
+  CHECK_VKR(m_pDriver, vkr);
+
+  vkr = ObjDisp(d)->BindBufferMemory(Unwrap(d), result.buf, result.mem, 0);
+  CHECK_VKR(m_pDriver, vkr);
+
+  return result;
+}
+
+bool VulkanAccelerationStructureManager::FixUpReplayBDAs(
+    VkAccelerationStructureInfo *asInfo, rdcarray<VkAccelerationStructureGeometryKHR> &geoms)
+{
+  RDCASSERT(asInfo);
+  RDCASSERT(asInfo->geometryData.size() == geoms.size());
+
+  const VkDevice d = m_pDriver->GetDev();
+
+  for(size_t i = 0; i < geoms.size(); ++i)
+  {
+    VkBuffer buf = asInfo->geometryData[i].replayBuf;
+    VkAccelerationStructureGeometryKHR &geom = geoms[i];
+
+    const VkBufferDeviceAddressInfo addrInfo = {VK_STRUCTURE_TYPE_BUFFER_DEVICE_ADDRESS_INFO, NULL,
+                                                Unwrap(buf)};
+    const VkDeviceAddress bufAddr = ObjDisp(d)->GetBufferDeviceAddressKHR(Unwrap(d), &addrInfo);
+
+    switch(geom.geometryType)
+    {
+      case VK_GEOMETRY_TYPE_TRIANGLES_KHR:
+      {
+        VkAccelerationStructureGeometryTrianglesDataKHR &tri = geom.geometry.triangles;
+
+        tri.vertexData.deviceAddress = bufAddr;
+
+        if(tri.indexType != VK_INDEX_TYPE_NONE_KHR)
+          tri.indexData.deviceAddress = bufAddr;
```

Comment:
I meant to leave a comment here but forgot while jumping around - should the vertex and index data both be pointing to the base of the buffer? shouldn't one be offset if indices are in use?

Reply:
Assuming I've read the spec right, the offsets are provided by the associated `VkAccelerationStructureBuildRangeInfoKHR`.  So if an index buffer is used it starts reading them from `primitiveOffset` bytes from this address, otherwise it consumes the vertices from this address directly.

Reply:
Ahh I see what you mean, yes you are right there is a (somewhat redundant?) offset internally which is set by CopyInputBuffers and propagate through which covers this.